File: Solutions/Cpp/merge_two_sorted_lists.cpp
/*
    [EASY]
    21. Merge Two Sorted Lists

    Concepts:
    - linked list

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 19.37 MB  [Beats 86.54%]
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* res = new ListNode();
        ListNode* head = res;

        while (list1 != nullptr && list2 != nullptr) {
            if (list1->val <= list2->val) {
                res->next = list1;
                list1 = list1->next;
            } else {
                res->next = list2;
                list2 = list2->next;
            }

            res = res->next;
        }

        if (list1 != nullptr)
            res->next = list1;
        else if (list2 != nullptr)
            res->next = list2;

        return head->next;
    }
};

File: Solutions/Rust/merge_two_sorted_lists.rs
/*
    [EASY]
    21. Merge Two Sorted Lists

    Concepts:
    - linked list

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.20 MB   [Beats 54.86%]
*/

// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(mut l1: Option<Box<ListNode>>, mut l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut head: Box<ListNode> = Box::new(ListNode::new(-1));
        let mut res: &mut Box<ListNode> = &mut head;

        while l1.is_some() && l2.is_some() {
            if let [Some(a), Some(b)] = [l1.as_ref(), l2.as_ref()] {
                if a.val <= b.val {
                    res.next = l1.clone();
                    l1 = l1.and_then(|node| node.next);
                } else {
                    res.next = l2.clone();
                    l2 = l2.and_then(|node| node.next);
                }
            }

            res = res.next.as_mut().unwrap();
        }

        if l1.is_some() {
            res.next = l1.clone();
        } else if l2.is_some() {
            res.next = l2.clone();
        }

        return head.next;
    }
}